File: src/simulation/space/tile/WindCenter.java
package simulation.space.tile;

import kotlin.Pair;
import simulation.space.SpaceData;
import simulation.space.WorldMap;
import simulation.space.resource.Resource;

import java.util.List;

import static simulation.Controller.session;

public class WindCenter {
    private Wind wind;
    private Wind _newWind;

     WindCenter() {
        wind = new Wind();
    }

    Wind getWind() {
        return wind;
    }

    void startUpdate() {
        _newWind = new Wind();
    }

    void useWind(List<Resource> resources) {
        for (Resource resource : resources) {
            if (!resource.getGenome().isMovable()) {
                continue;
            }
            double overallWindLevel = wind.affectedTiles.stream()
                    .map(Pair::getSecond)
                    .reduce(Double::sum)
                    .orElse(0.0);
            for (Pair<Tile, Double> pair : wind.affectedTiles) {
                int part = (int) (resource.getAmount() * pair.getSecond() / overallWindLevel *
                        Math.min(pair.getSecond() * 0.0001 / resource.getGenome().getMass(), 1));
                if (part > 0) {
                    pair.getFirst().addDelayedResource(resource.getCleanPart(part));
                }
            }
        }
    }

    void middleUpdate(int x, int y) {
        WorldMap map = session.world.getMap();
        Tile host = map.get(x, y);
        host.getNeighbours().forEach(n -> setWindByTemperature(n, host));

        if (!_newWind.isStill())
            return;

        propagateWindStraight(map.get(x - 1, y), map.get(x + 1, y), host);
        propagateWindStraight(map.get(x + 1, y), map.get(x - 1, y), host);
        propagateWindStraight(map.get(x, y - 1), map.get(x, y + 1), host);
        propagateWindStraight(map.get(x, y + 1), map.get(x, y - 1), host);

        if (!_newWind.isStill()) {//TODO better to addAll wind for cross tiles than try to fetch it; cut wind on large level changes
            return;
        }

        propagateWindFillIn(map.get(x - 1, y), map.get(x - 2, y));
        propagateWindFillIn(map.get(x + 1, y), map.get(x + 2, y));
        propagateWindFillIn(map.get(x, y - 1), map.get(x, y - 2));
        propagateWindFillIn(map.get(x, y + 1), map.get(x, y + 2));
    }

    public void finishUpdate() {
        wind = _newWind;
    }

    private void setWindByTemperature(Tile tile, Tile master) {
        int temperatureChange = SpaceData.INSTANCE.getData().getTemperatureToWindCoefficient();
        if (tile.getLevel() + 2 < master.getLevel()) {
            temperatureChange *= 5;
        }
        if (tile != null) {
            double level = ((double) tile.getTemperature() - 1 - master.getTemperature()) / temperatureChange;
            if (level > 0) {
                _newWind.changeLevelOnTile(tile, level);
            }
        }
    }

    private void propagateWindStraight(Tile target, Tile tile, Tile master) {
        if (tile != null && target != null) {
            double level = tile.getWind().getPureLevelByTile(master) - SpaceData.INSTANCE.getData().getWindPropagation();
            if (level > 0) {
                _newWind.changeLevelOnTile(target, level);
            }
        }
    }

    private void propagateWindFillIn(Tile tile, Tile target) {
        if (tile != null && target != null) {
            double level = tile.getWind().getLevelByTile(target) - session.windFillIn;
            if (level > 0) {
                _newWind.isFilling = true;
                _newWind.changeLevelOnTile(tile, level);
            }
        }
    }
}

File: EcosystemSimulation/src/shmp/simulation/space/tile/WindCenter.kt
package shmp.simulation.space.tile

import shmp.simulation.space.SpaceData.data
import shmp.simulation.space.WorldMap
import shmp.simulation.space.resource.Resource
import shmp.simulation.space.resource.Taker
import shmp.simulation.space.resource.freeMarker
import kotlin.math.max
import kotlin.math.pow


class WindCenter internal constructor() {
    var wind = Wind()
        private set

    private var _newWind = Wind()

    fun startUpdate() {
        _newWind = Wind()
    }

    fun useWind(resources: List<Resource>) {
        for (resource in resources) {
            if (!resource.genome.isMovable)
                continue

            for ((tile, t) in wind.affectedTiles) {
                val part = (resource.amount * t.pow(1) / wind.sumLevel * getFlyCoefficient(resource)).toInt()

                if (part > 0)
                    tile.addDelayedResource(resource.getCleanPart(part, Taker.WindTaker))
            }
        }
    }

    private fun getFlyCoefficient(resource: Resource) = 0.0001 /
            resource.genome.mass /
            if (resource.core.ownershipMarker == freeMarker) 1 else 10

    fun middleUpdate(x: Int, y: Int, map: WorldMap) {
        val host = map[x, y]
                ?: return

        host.neighbours.forEach { setWindByTemperature(it, host) }

        if (!_newWind.isStill)
            return

        propagateWindStraight(map[x - 1, y], map[x + 1, y], host)
        propagateWindStraight(map[x + 1, y], map[x - 1, y], host)
        propagateWindStraight(map[x, y - 1], map[x, y + 1], host)
        propagateWindStraight(map[x, y + 1], map[x, y - 1], host)
        //TODO better to addAll wind for cross tiles than try to fetch it; cut wind on large level changes
        if (!_newWind.isStill)
            return
        propagateWindFillIn(map[x - 1, y], map[x - 2, y])
        propagateWindFillIn(map[x + 1, y], map[x + 2, y])
        propagateWindFillIn(map[x, y - 1], map[x, y - 2])
        propagateWindFillIn(map[x, y + 1], map[x, y + 2])
    }

    fun finishUpdate() {
        wind = _newWind
    }

    private fun setWindByTemperature(tile: Tile?, master: Tile) {
        tile ?: return

        var change = data.temperatureToWindCoefficient
        if (tile.level + 2 < master.level)
            change *= 5
        if (tile.type == master.type)
            change *= 5

        val level = max(tile.temperature.toDouble() - 1 - master.temperature, 0.0) / change
        if (level > 0)
            _newWind.changeLevelOnTile(tile, level)
    }

    private fun propagateWindStraight(target: Tile?, tile: Tile?, master: Tile) {
        tile ?: return
        target ?: return

        val level = tile.wind.getPureLevelByTile(master) - data.windPropagation
        if (level > 0)
            _newWind.changeLevelOnTile(target, level)
    }

    private fun propagateWindFillIn(tile: Tile?, target: Tile?) {
        tile ?: return
        target ?: return

        val level = tile.wind.getLevelByTile(target) - data.windFillIn
        if (level > 0) {
            _newWind.isFilling = true
            _newWind.changeLevelOnTile(tile, level)
        }
    }
}
